File: pages/api/sync-activities.js
/**
 * /api/sync-activities — POST
 *
 * Triggers an on-demand SFDC Tasks → Postgres activities sync
 * (last 30 days, subtypes: Call, Email, LinkedInMessage, Task).
 *
 * Returns: { synced: N, errors: M, completedAt: ISO_timestamp }
 * Also writes to sync_log on completion.
 */

import { query } from '../../lib/db';
import jsforce from 'jsforce';

// ── SFDC credentials ─────────────────────────────────────────────────────────
const SFDC_USER     = process.env.SFDC_USERNAME  || 'gray.hoffman@getathelas.com';
const SFDC_PASSWORD = process.env.SFDC_PASSWORD  || 'ctk0WZK*rzw@tyh!pnp';
const SFDC_TOKEN    = process.env.SFDC_TOKEN     || 'zK9vAeYocFwweG6zBmKDvO2F';
const SFDC_INSTANCE = process.env.SFDC_INSTANCE  || 'https://athelas.my.salesforce.com';

const SOQL = `
  SELECT Id, Subject, ActivityDate, TaskSubtype, CallType,
         CallDurationInSeconds, CallDisposition, Description,
         Status, WhoId, WhatId, Who.Name, What.Name,
         OwnerId, Owner.Name, CreatedDate, LastModifiedDate
  FROM Task
  WHERE ActivityDate >= LAST_N_DAYS:30
  AND TaskSubtype IN ('Call', 'Email', 'LinkedInMessage', 'Task')
  ORDER BY ActivityDate DESC
  LIMIT 5000
`;

function mapSubtype(subtype, subject) {
  if (subtype === 'Call')            return 'call';
  if (subtype === 'Email')           return 'email';
  if (subtype === 'LinkedInMessage') return 'linkedin';
  if (subtype === 'Task')            return 'task';
  const sl = (subject || '').toLowerCase();
  if (sl.includes('call'))     return 'call';
  if (sl.includes('email'))    return 'email';
  if (sl.includes('linkedin')) return 'linkedin';
  return 'task';
}

function cors(res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
}

export default async function handler(req, res) {
  cors(res);
  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'POST')   return res.status(405).json({ error: 'Method not allowed. Use POST.' });

  const startedAt = new Date();
  let synced = 0;
  let errors = 0;

  try {
    // 1. Connect to SFDC via jsforce
    const conn = new jsforce.Connection({ loginUrl: SFDC_INSTANCE });
    await conn.login(SFDC_USER, SFDC_PASSWORD + SFDC_TOKEN);

    // 2. Query tasks
    const result = await conn.query(SOQL);
    let records = result.records || [];

    // Handle queryMore if needed
    let queryResult = result;
    while (!queryResult.done && queryResult.nextRecordsUrl) {
      queryResult = await conn.queryMore(queryResult.nextRecordsUrl);
      records = records.concat(queryResult.records || []);
    }

    // 3. Build account/contact sfdc_id lookup sets
    const acctRes = await query('SELECT sfdc_id FROM accounts WHERE sfdc_id IS NOT NULL');
    const contRes = await query('SELECT sfdc_id FROM contacts WHERE sfdc_id IS NOT NULL');
    const accountIds = new Set(acctRes.rows.map(r => r.sfdc_id));
    const contactIds = new Set(contRes.rows.map(r => r.sfdc_id));

    // 4. Upsert in batches
    const UPSERT_SQL = `
      INSERT INTO activities (
        sfdc_id, account_sfdc_id, contact_sfdc_id,
        type, subject, activity_date,
        outcome, notes, rep,
        source_system, call_duration_seconds, call_disposition,
        created_at
      ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,'sfdc',$10,$11,NOW())
      ON CONFLICT (sfdc_id) DO UPDATE SET
        account_sfdc_id       = EXCLUDED.account_sfdc_id,
        contact_sfdc_id       = EXCLUDED.contact_sfdc_id,
        type                  = EXCLUDED.type,
        subject               = EXCLUDED.subject,
        activity_date         = EXCLUDED.activity_date,
        outcome               = EXCLUDED.outcome,
        notes                 = EXCLUDED.notes,
        rep                   = EXCLUDED.rep,
        call_duration_seconds = EXCLUDED.call_duration_seconds,
        call_disposition      = EXCLUDED.call_disposition,
        source_system         = 'sfdc'
    `;

    for (const rec of records) {
      try {
        const whatId      = rec.WhatId || null;
        const whoId       = rec.WhoId  || null;
        const actDate     = rec.ActivityDate ? new Date(rec.ActivityDate) : null;
        const ownerName   = rec.Owner?.Name || null;
        const acctSfdcId  = whatId && accountIds.has(whatId) ? whatId : null;
        const contSfdcId  = whoId  && contactIds.has(whoId)  ? whoId  : null;
        const duration    = rec.CallDurationInSeconds != null ? parseInt(rec.CallDurationInSeconds) : null;

        await query(UPSERT_SQL, [
          rec.Id,
          acctSfdcId,
          contSfdcId,
          mapSubtype(rec.TaskSubtype, rec.Subject),
          (rec.Subject || '').slice(0, 500),
          actDate,
          rec.CallDisposition || null,
          rec.Description ? rec.Description.slice(0, 2000) : null,
          ownerName,
          duration,
          rec.CallDisposition || null,
        ]);
        synced++;
      } catch (rowErr) {
        console.error(`[sync-activities] row error ${rec.Id}:`, rowErr.message);
        errors++;
      }
    }

    // 5. Write sync_log
    const completedAt = new Date();
    await query(
      `INSERT INTO sync_log (table_name, sync_type, completed_at, records_synced, errors, started_at)
       VALUES ('activities', 'incremental_30d', $1, $2, $3, $4)`,
      [completedAt, synced, errors, startedAt]
    );

    return res.status(200).json({
      synced,
      errors,
      total: records.length,
      completedAt: completedAt.toISOString(),
    });

  } catch (err) {
    console.error('[sync-activities] fatal error:', err);
    // Still try to log the failure
    try {
      await query(
        `INSERT INTO sync_log (table_name, sync_type, completed_at, records_synced, errors, started_at, notes)
         VALUES ('activities', 'incremental_30d', NOW(), $1, $2, $3, $4)`,
        [synced, errors + 1, startedAt, err.message]
      );
    } catch (_) { /* ignore log failure */ }

    return res.status(500).json({
      error: err.message,
      synced,
      errors: errors + 1,
      completedAt: new Date().toISOString(),
    });
  }
}

File: scripts/sync_activities.py
#!/usr/bin/env python3
"""
sync_activities.py - Sync SFDC Tasks to Postgres activities table
Pulls last 30 days of Tasks (Call, Email, LinkedInMessage, Task subtypes)
and upserts into the activities table.
"""

import os
import sys
import time
from datetime import datetime, timezone

import psycopg2
import psycopg2.extras
from simple_salesforce import Salesforce, SalesforceAuthenticationFailed

# ── Config ──────────────────────────────────────────────────────────────────
DB_URL  = os.environ.get(
    "DATABASE_URL",
    "postgresql://neondb_owner:npg_zr6DfhL0gelj@ep-bold-leaf-an92homd.c-6.us-east-1.aws.neon.tech/neondb?sslmode=require"
)
SFDC_USER     = "gray.hoffman@getathelas.com"
SFDC_PASSWORD = "ctk0WZK*rzw@tyh!pnp"
SFDC_TOKEN    = "zK9vAeYocFwweG6zBmKDvO2F"
SFDC_INSTANCE = "https://athelas.my.salesforce.com"

SOQL = """
SELECT Id, Subject, ActivityDate, TaskSubtype, CallType,
       CallDurationInSeconds, CallDisposition, Description,
       Status, WhoId, WhatId, Who.Name, What.Name,
       OwnerId, Owner.Name, CreatedDate, LastModifiedDate
FROM Task
WHERE ActivityDate >= LAST_N_DAYS:30
AND TaskSubtype IN ('Call', 'Email', 'LinkedInMessage', 'Task')
ORDER BY ActivityDate DESC
LIMIT 5000
"""

# ── Helpers ──────────────────────────────────────────────────────────────────

def map_subtype(subtype, subject):
    """Map SFDC TaskSubtype to our type values."""
    if subtype in ("Call",):
        return "call"
    if subtype in ("Email",):
        return "email"
    if subtype in ("LinkedInMessage",):
        return "linkedin"
    if subtype in ("Task",):
        return "task"
    # Fallback: infer from subject
    if subject:
        sl = subject.lower()
        if "call" in sl:
            return "call"
        if "email" in sl:
            return "email"
        if "linkedin" in sl or "li" in sl:
            return "linkedin"
    return "task"


def build_account_index(cur):
    """Return {sfdc_id: sfdc_id} mapping for accounts (sfdc_id → sfdc_id, just for lookup)."""
    cur.execute("SELECT sfdc_id FROM accounts WHERE sfdc_id IS NOT NULL")
    return {row[0] for row in cur.fetchall()}


def build_contact_index(cur):
    """Return set of known contact sfdc_ids."""
    cur.execute("SELECT sfdc_id FROM contacts WHERE sfdc_id IS NOT NULL")
    return {row[0] for row in cur.fetchall()}


# ── Main ─────────────────────────────────────────────────────────────────────

def run_sync():
    started = datetime.now(timezone.utc)
    print(f"[{started.isoformat()}] Starting SFDC → Postgres activities sync…")

    # 1. Connect to SFDC
    print("  Connecting to Salesforce…")
    try:
        sf = Salesforce(
            username=SFDC_USER,
            password=SFDC_PASSWORD,
            security_token=SFDC_TOKEN,
            instance_url=SFDC_INSTANCE,
        )
        print("  ✓ Salesforce connected")
    except SalesforceAuthenticationFailed as e:
        print(f"  ✗ SFDC auth failed: {e}")
        sys.exit(1)

    # 2. Query SFDC
    print("  Querying Tasks (last 30 days)…")
    result = sf.query_all(SOQL)
    records = result["records"]
    print(f"  ✓ Got {len(records)} records from SFDC")

    if not records:
        print("  No records to sync. Done.")
        return

    # 3. Connect to Postgres
    print("  Connecting to Postgres…")
    conn = psycopg2.connect(DB_URL)
    cur  = conn.cursor()
    print("  ✓ Postgres connected")

    # 4. Build lookup indexes
    account_sfdc_ids = build_account_index(cur)
    contact_sfdc_ids = build_contact_index(cur)
    print(f"  Account index: {len(account_sfdc_ids)} entries")
    print(f"  Contact index: {len(contact_sfdc_ids)} entries")

    # 5. Upsert
    UPSERT_SQL = """
        INSERT INTO activities (
            sfdc_id, account_sfdc_id, contact_sfdc_id,
            type, subject, activity_date,
            outcome, notes, rep,
            source_system, call_duration_seconds,
            call_disposition, created_at
        ) VALUES (
            %(sfdc_id)s, %(account_sfdc_id)s, %(contact_sfdc_id)s,
            %(type)s, %(subject)s, %(activity_date)s,
            %(outcome)s, %(notes)s, %(rep)s,
            'sfdc', %(call_duration_seconds)s,
            %(call_disposition)s, NOW()
        )
        ON CONFLICT (sfdc_id) DO UPDATE SET
            account_sfdc_id     = EXCLUDED.account_sfdc_id,
            contact_sfdc_id     = EXCLUDED.contact_sfdc_id,
            type                = EXCLUDED.type,
            subject             = EXCLUDED.subject,
            activity_date       = EXCLUDED.activity_date,
            outcome             = EXCLUDED.outcome,
            notes               = EXCLUDED.notes,
            rep                 = EXCLUDED.rep,
            call_duration_seconds = EXCLUDED.call_duration_seconds,
            call_disposition    = EXCLUDED.call_disposition,
            source_system       = 'sfdc'
    """

    synced = 0
    errors = 0
    batch  = []

    for rec in records:
        try:
            what_id = rec.get("WhatId")
            who_id  = rec.get("WhoId")

            # Only link if we have the account/contact in our DB
            account_sfdc_id = what_id if what_id and what_id in account_sfdc_ids else None
            contact_sfdc_id = who_id  if who_id  and who_id  in contact_sfdc_ids else None

            # Parse ActivityDate (can be date string or None)
            act_date = rec.get("ActivityDate")
            if act_date and isinstance(act_date, str):
                try:
                    act_date = datetime.strptime(act_date, "%Y-%m-%d")
                except ValueError:
                    act_date = None

            # Owner.Name via relationship
            owner_name = None
            owner_rel  = rec.get("Owner")
            if owner_rel and isinstance(owner_rel, dict):
                owner_name = owner_rel.get("Name")

            row = {
                "sfdc_id":               rec["Id"],
                "account_sfdc_id":       account_sfdc_id,
                "contact_sfdc_id":       contact_sfdc_id,
                "type":                  map_subtype(rec.get("TaskSubtype"), rec.get("Subject")),
                "subject":               (rec.get("Subject") or "")[:500],
                "activity_date":         act_date,
                "outcome":               rec.get("CallDisposition"),
                "notes":                 (rec.get("Description") or "")[:2000] if rec.get("Description") else None,
                "rep":                   owner_name,
                "call_duration_seconds": rec.get("CallDurationInSeconds"),
                "call_disposition":      rec.get("CallDisposition"),
            }
            batch.append(row)

        except Exception as e:
            print(f"  ✗ Row error for {rec.get('Id', '?')}: {e}")
            errors += 1
            continue

        # Flush every 500
        if len(batch) >= 500:
            try:
                psycopg2.extras.execute_batch(cur, UPSERT_SQL, batch, page_size=500)
                conn.commit()
                synced += len(batch)
                print(f"  … {synced} upserted so far…")
                batch = []
            except Exception as e:
                print(f"  ✗ Batch error: {e}")
                conn.rollback()
                errors += len(batch)
                batch = []

    # Final flush
    if batch:
        try:
            psycopg2.extras.execute_batch(cur, UPSERT_SQL, batch, page_size=500)
            conn.commit()
            synced += len(batch)
        except Exception as e:
            print(f"  ✗ Final batch error: {e}")
            conn.rollback()
            errors += len(batch)

    # 6. Write sync_log
    completed = datetime.now(timezone.utc)
    try:
        cur.execute(
            """
            INSERT INTO sync_log (table_name, sync_type, completed_at, records_synced, errors, started_at)
            VALUES ('activities', 'incremental_30d', %s, %s, %s, %s)
            """,
            (completed, synced, errors, started)
        )
        conn.commit()
        print(f"  ✓ sync_log written")
    except Exception as e:
        print(f"  ✗ sync_log error: {e}")
        conn.rollback()

    cur.close()
    conn.close()

    elapsed = (completed - started).total_seconds()
    print(f"\n✅ Sync complete: {synced} upserted, {errors} errors ({elapsed:.1f}s)")
    return synced, errors


if __name__ == "__main__":
    run_sync()
